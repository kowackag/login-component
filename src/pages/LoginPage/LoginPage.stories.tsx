import { Meta, StoryObj } from "@storybook/react";
import  LoginPage  from "./LoginPage";

const meta: Meta<typeof LoginPage> = {
	component: LoginPage,
	render: () => {
		
		return <LoginPage  />;
	},
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
	args: {},
};
