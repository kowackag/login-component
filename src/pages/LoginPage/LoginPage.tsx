import { FieldErrors, UseFormRegister, useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";

import { LoginPageTemplate } from "@/components/templates/LoginPageTemplate/LoginPageTemplate";

import { schema } from "@/pages/authSchema";

type LoginFieldsTypes = { email: string; password: string; toSave: boolean };
export type LoginFormType = {
	onSubmit: React.FormEventHandler;
	errors?: FieldErrors<LoginFieldsTypes>;
	register: UseFormRegister<LoginFieldsTypes>;
};

type LoginFormDataTypes = {
	email: string;
	password: string;
	toSave: boolean;
};
function LoginPage() {
	const {
		register,
		formState: { errors },
		handleSubmit,
	} = useForm<LoginFormDataTypes>({
		resolver: zodResolver(schema),
		defaultValues: {
			email: "",
			password: "",
			toSave: false,
		},
	});

	const onSubmit = handleSubmit((data) => {
		console.log(data);
	});

	return <LoginPageTemplate onSubmit={onSubmit} register={register} errors={errors} />;
}

export default LoginPage;
