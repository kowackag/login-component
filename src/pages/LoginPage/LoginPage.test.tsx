import { expect, it, describe } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import LoginPage from "./LoginPage";
import { BrowserRouter } from "react-router-dom";

function flushPromises() {
	return new Promise((resolve) => setImmediate(resolve));
}

describe("LoginPage", () => {
	it("component exists", () => {
		const wrapper = render(
			<BrowserRouter>
				<LoginPage />
			</BrowserRouter>,
		);
		expect(wrapper).toBeTruthy();
		expect(wrapper.container).toBeInTheDocument();
	});

	it("render appropriate title and form", () => {
		render(
			<BrowserRouter>
				<LoginPage />
			</BrowserRouter>,
		);
		const form = screen.getByTestId("login-form");
		expect(form).toBeInTheDocument();
		const title = screen.getByTestId("title");
		expect(title).toHaveRole("heading");
		expect(title).toHaveTextContent("Login");
		expect(title).toBeInTheDocument();
	});
	it("renders correct errors when email/password does not meet requirements", async () => {
		render(
			<BrowserRouter>
				<LoginPage />
			</BrowserRouter>,
		);
		const button = screen.getByTestId("btn-submit");
		expect(button).toHaveTextContent("Login");

		fireEvent.click(button);

		const errorList = await screen.findAllByTestId("error");
		expect(errorList).toHaveLength(2);
		expect(errorList[0]).toHaveTextContent("Invalid email address");
		expect(errorList[1]).toHaveTextContent("Invalid password");
	});
	it("hides error after filling form correctly", async () => {
		render(
			<BrowserRouter>
				<LoginPage />
			</BrowserRouter>,
		);
		const button = screen.getByTestId("btn-submit");
		const passwordInput = screen.getByLabelText("Password") as HTMLInputElement;

		expect(button).toHaveTextContent("Login");

		fireEvent.click(button);

		const errorList = await screen.findAllByTestId("error");
		expect(errorList).toHaveLength(2);

		fireEvent.change(passwordInput, { target: { value: "own-password" } });

		expect(passwordInput.value).toBe("own-password");

		fireEvent.click(button);
		await flushPromises();
		const newErrorList = await screen.findAllByTestId("error");
		expect(newErrorList).toHaveLength(1);
	});

	it("is reset after processing form", async () => {
		const { container } = render(
			<BrowserRouter>
				<LoginPage />
			</BrowserRouter>,
		);
		const button = screen.getByTestId("btn-submit");
		expect(button).toHaveTextContent("Login");

		const passwordInput = container.querySelector("#password") as HTMLInputElement;

		fireEvent.change(passwordInput, { target: { value: "own-password" } });

		const emailInput = container.querySelector("#email") as HTMLInputElement;
		fireEvent.change(emailInput, { target: { value: "someone@wp.pl" } });

		expect(passwordInput.value).toBe("own-password");
		expect(emailInput.value).toBe("someone@wp.pl");

		fireEvent.click(button);

		expect(passwordInput.innerHTML).toBe("");
		expect(emailInput.innerHTML).toBe("");
	});

	it("renders correctly", () => {
		const { container } = render(
			<BrowserRouter>
				<LoginPage />
			</BrowserRouter>,
		);
		expect(container).toMatchSnapshot();
	});
});
