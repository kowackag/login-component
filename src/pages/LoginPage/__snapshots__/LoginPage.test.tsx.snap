// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > renders correctly 1`] = `
<div>
  <div
    class="flex min-h-screen w-full bg-navy-blue"
  >
    <section
      class="m-auto min-h-[556px] w-4/5 rounded-lg bg-white px-11 py-10 font-primaryRegular shadow-md-2 sm:w-96"
    >
      <h1
        class="mb-4 text-lg font-bold uppercase"
        data-testid="title"
      >
        Login
      </h1>
      <form
        data-testid="login-form"
      >
        <div
          class="relative py-3"
        >
          <label
            class="inline-block font-primaryMedium"
            for="email"
          >
            Email
          </label>
          <input
            class="w-full rounded-md border-2 border-neutral-300 bg-transparent px-2 leading-9 outline-none"
            data-testid="email-input"
            id="email"
            name="email"
            type="text"
          />
        </div>
        <div
          class="relative py-3"
        >
          <label
            class="inline-block font-primaryMedium"
            for="password"
          >
            Password
          </label>
          <input
            class="w-full rounded-md border-2 border-neutral-300 bg-transparent px-2 leading-9 outline-none"
            data-testid="password-input"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <div
          class="justify-beetwen mb-4 flex align-middle"
        >
          <input
            class="peer fixed h-5 w-5 opacity-0"
            data-testid="checkbox-toSave"
            id="toSave"
            name="toSave"
            type="checkbox"
          />
          <p
            class="mr-3 inline-block h-5 w-5 rounded-sm border border-neutral-300 p-0 leading-5 text-white peer-checked:border-none peer-checked:bg-bright-rose "
          >
            <svg
              height="16px"
              width="16px"
            >
              <use
                href="/icons.svg#check"
              />
            </svg>
          </p>
          <label
            class="font-primaryMedium leading-5"
            for="toSave"
          >
            Remember me?
          </label>
        </div>
        <button
          class="my-2 w-full rounded-lg bg-bright-rose text-center font-primaryMedium text-sm uppercase leading-10 tracking-widest text-neutral-100 shadow-btn"
          data-testid="btn-submit"
          type="submit"
        >
          Login
        </button>
      </form>
      <button
        class="w-full text-end font-primaryRegular text-neutral-400"
        data-testid="btn-restore-password"
      >
        Forgot password?
      </button>
      <div
        class="my-3 flex items-center"
      >
        <span
          class="flex-grow border border-b border-b-gray-300"
          data-testid="hrLine"
        />
        <p
          class="rounded-md border-2  border-neutral-300 px-1 py-0.5 font-primaryRegular uppercase text-neutral-400"
        >
          or
        </p>
        <span
          class="flex-grow border border-b border-b-gray-300"
          data-testid="hrLine"
        />
      </div>
      <div
        class="flex justify-center"
        data-testid="icon-box"
      >
        <a
          aria-label="google"
          class="mx-2 my-2"
          data-testid="link"
          href="https://mail.google.com/"
          target="_blank"
        >
          <svg
            height="36px"
            width="36px"
          >
            <use
              href="/google.svg#google"
            />
          </svg>
        </a>
        <a
          aria-label="facebook"
          class="mx-2 my-2"
          data-testid="link"
          href="https://www.facebook.com/"
          target="_blank"
        >
          <svg
            height="36px"
            width="36px"
          >
            <use
              href="/facebook.svg#facebook"
            />
          </svg>
        </a>
        <a
          aria-label="linkedin"
          class="mx-2 my-2"
          data-testid="link"
          href="https://www.linkedin.com/in/malgorzata-kowacka/"
          target="_blank"
        >
          <svg
            height="36px"
            width="36px"
          >
            <use
              href="/linkedin.svg#linkedin"
            />
          </svg>
        </a>
      </div>
      <div
        class="mb-0 mt-4 flex justify-center font-primaryMedium"
      >
        <p>
          Need an account?
        </p>
        <button
          class="ml-1 uppercase underline"
          data-testid="btn-navigate"
        >
          sign up
        </button>
      </div>
    </section>
  </div>
</div>
`;
